### PR TITLE
Standardizes object deconstruction  throughout the codebase. 

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
When it comes to deconstructing an object we have `proc/deconstruct()` & `NO_DECONSTRUCT`

Lets talk about the flag first. 

**Problems with `NO_DECONSTRUCTION`**
I know what the comment says on what it should do
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/__DEFINES/obj_flags.dm#L18

But everywhere people have decided to give their own meaning/definition to this flag. Here are some examples on how this flag is used

**1. Make the object just disappear(not drop anything) when deconstructed**
This is by far the largest use case everywhere. If an object is deconstructed(either via tools or smashed apart) then if it has this flag it should not drop any of its contents but just disappear. You have seen this code pattern used everywhere
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/game/machinery/constructable_frame.dm#L26-L31

This behaviour is then leveraged by 2 important components.

When an object is frozen, if it is deconstructed it should just disappear without leaving any traces behind
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/datums/elements/frozen.dm#L66-L67

By hologram objects. Obviously if you destroy an hologram nothing real should drop out
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/modules/holodeck/computer.dm#L301-L304

And there are other use cases as well but we won't go into them as they aren't as significant as these.

**2. To stop an object from being wrenched ??**
Yeah this one is weird. Like why? I understand in some instances (chair, table, rack etc) a wrench can be used to deconstruct a object so using the flag there to stop it from happening makes sense but why can't we even anchor an object just because of this flag?
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/game/objects/objs.dm#L368-L369
This is one of those instances where somebody just decided this behaviour for their own convenience just like the above example with no explanation as to why

**3. To stop using tools to deconstruct the object** 
This was the original intent of the flag but it is enforced in few places far & between. One example is when deconstructing the a machine via crowbar.
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/game/machinery/_machinery.dm#L811

But machines are a special dual use case for this flag. Because if you look at its deconstruct proc the flag also prevents the machine from spawning a frame.
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/game/machinery/_machinery.dm#L820-L822

How can 1 flag serve 2 purposes within the same type?

**4. Simply forget to check for this flag altogether**
Yup if you find this flag not doing its job for some objects don't be surprised. People & sometimes even maintainers just forget that it even exists
https://github.com/tgstation/tgstation/blob/b5593bc6930cb60803214869a7b94c84e7baa02c/code/game/objects/items/piggy_bank.dm#L66-L67

**Solution**
These are the main examples i found. As you can see the same flag can perform 2 different functions within the same type and do something else in a different object & in some instances don't even work cause people just forget, etc. 

In order to bring consistency to this flag we need to move it to the atom level where it means the same thing everywhere. Where in the atom you may ask? .Well, I'll just post what MrMelbert said in https://github.com/tgstation/tgstation/pull/81656#discussion_r1503086862

> ...Ideally the .deconstruct call would handle NO_DECONSTRUCTION handling as it wants,

Yup that's the ideal case now. This flag is checked directly in `deconstruct()`. Now like i said we want to give a universal definition to this flag and as you have seen from my examples it is used in 3 cases
1) Make an object disappear(doesn't dropping anything) when deconstructed
2) Stop it from being wrenched
3) Stop it from being deconstructed via tools

We can't enforce points 2 & 3 inside `deconstruct()` which leaves us with only case 1) i.e. make the object disappear.  And that's what i have done. Therefore after more than a decade or since this flag got introduced `NO_DECONSTRUCT` now has a new definition as of 2024

_"Make an object disappear(don't dropping anything) when deconstructed either via tools or forcefully smashed apart"_

Now i very well understand this will open up bugs in places where cases 2 & 3 are required but its worth it. In fact they could even be qol changes for all we know so who knows it might even benefit us but for now we need to give a universal definition to this flag to bring some consistency & that's what this PR does.

**Problem with deconstruct()**
This proc actually sends out a signal which is currently used by the material container but could be used by other objects later on.
https://github.com/tgstation/tgstation/blob/3e84c3e6dad33c831ac259f52f2f023680e4899b/code/game/objects/obj_defense.dm#L160

So objects that override this proc should call its parent. Sadly that isn't the case in many instances like such
https://github.com/tgstation/tgstation/blob/3e84c3e6dad33c831ac259f52f2f023680e4899b/code/game/machinery/deployable.dm#L20-L23

Instead of `return ..()` which would delete the object & send the signal it deletes the object directly thus the signal never gets sent.

**Solution**
Make this proc non overridable. For objects to add their own custom deconstruction behaviour a new proc has been introduced `atom_deconstruct()` Subtypes should now override this proc to handle object deconstruction.

If objects have certain important stuff inside them (like mobs in machines for example) they want to drop by handling `NO_DECONSTRUCT` flag in a more carefully customized way they can do this by overriding `handle_deconstruct()` which by default delegates to `atom_deconstruct()` if the `NO_DECONSTRUCT` flag is absent. This proc will allow you to handle the flag in a more customized way if you ever need to.

## Why It's Good For The Game
1) I'm goanna post the full comment from MrMelbert https://github.com/tgstation/tgstation/pull/81656#discussion_r1503086862

> ...Ideally the .deconstruct call would handle NO_DECONSTRUCTION handling as it wants, but there's a shocking lack of consistency around NO_DECONSTRUCTION, where some objects treat it as "allow deconstruction, but make it drop no parts" and others simply "disallow deconstruction at all"

  This PR now makes `NO_DECONSTRUCTION` handled by `deconstruct()` & gives this flag the consistency it deserves. Not to mention as shown in case 4 there are objects that simply forgot to check for this flag. Now it applies for those missing instances as well.

2) No more copying pasting the most overused code pattern in this code base history `if(obj_flags & NO_DECONSTRUCTION)`. Just makes code cleaner everywhere

3) All objects now send the `COMSIG_OBJ_DECONSTRUCT` signal on object deconstruction which is now available for use should you need it

## Changelog
:cl:
refactor: refactors how objects are deconstructed in relation to the `NO_DECONSTRUCTION` flag. Certain objects & machinery may display different tool interactions & behaviours when destroyed/deconstructed. Report these changes if you feel like they are bugs
/:cl:

### CODE REVIEW

**Comment by Admiral-Gigaton** [comment]
I aint readin all that

File: code/modules/power/pipecleaners.dm [hunk 1]
```diff
@@ -106,15 +106,13 @@ By design, d1 is the smallest direction and d2 is the highest
...
-	if(!(obj_flags & NO_DECONSTRUCTION))
-		var/turf/T = get_turf(loc)
-		if(T)
-			stored.forceMove(T)
-			stored = null
-		else
-			qdel(stored)
-	qdel(src)
+/obj/structure/pipe_cleaner/atom_deconstruct(disassembled = TRUE)
+	var/turf/T = get_turf(loc)
```
Comment (san7890) [inline code/modules/power/pipecleaners.dm:110]: yeaj

**Review by san7890 — CHANGES_REQUESTED** [review]
a lot of nits (most of them are pretty easy to fix, just need some time)- but there are a few important code considerations to look at.

**Comment by san7890** [comment]
also remove the `code:` line from your changelog it's not useful to players and coders will likely not see it to care for it

File: code/game/turfs/closed/walls.dm [hunk 2]
```diff
@@ -329,6 +329,8 @@
 /turf/closed/wall/rcd_vals(mob/user, obj/item/construction/rcd/the_rcd)
 	switch(the_rcd.mode)
 		if(RCD_DECONSTRUCT)
+			if(resistance_flags && INDESTRUCTIBLE)
+				return NONE
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/game/turfs/closed/walls.dm:333]: Phew,  that's a really good catch, thanks

**Comment by carshalash** [comment]
PLEASE test your PRs, i'm begging you.

File: code/game/objects/structures/tables_racks.dm [hunk 3]
```diff
@@ -297,20 +297,14 @@
...
-		if(buildstack)
-			new buildstack(T, buildstackamount)
-		else
-			for(var/i in custom_materials)
-				var/datum/material/M = i
-				new M.sheet_type(T, FLOOR(custom_materials[M] / SHEET_MATERIAL_AMOUNT, 1))
-		if(!wrench_disassembly)
-			new frame(T)
-		else
-			new framestack(T, framestackamount)
```
Comment (Dman757) [inline code/game/objects/structures/tables_racks.dm:312]: Disassembling a table no longer creates a frame if disassembled by a screw driver.

Also disassembling with a wrench doesn't produce iron rods

Comment (Dman757) [reply code/game/objects/structures/tables_racks.dm:312]: https://github.com/tgstation/tgstation/issues/82503

**Comment by SyncIt21** [comment]
> Problem is now those feedback messages are going to be displayed to users telling them that they are deconstructing something, where they actually aren't able to. And then nothing will happen.

NO_DECONSTRUCTION will no longer stop an object from being deconstructed. It will only make it dissapear after it has been deconstructed so the feedback messages still make sense because it still informs you of the action you are currently performing I.e deconstruction which you can still do because this flag no longer stops you from doing that.

If the flag is still being used to stop an objects deconstruction like before then it is incorrect now, and your goanna have to manually edit those atoms tool procs individually or create your own unique flag to stop that from happeningbecausethat is no longerthe purposeof this flag.

The example you have shown in your screenshot has no problems with it. You can still use your wrench to deconstruct the fan even though it will just dissapear afterwards, this is the new normal behavior now, you can either be ok with it(which I find acceptable because the flag does it's job) or make your own flag or something which would stop an object from being deconstructed.

It's a Matter of preference because think about it for hologram items, We can either be ok not allowing the items from being deconstructed(which makes sense cause you can't use real tools on holograms) or we can allow the hologram to be deconstructed but make it dissapear(which also makes sense and could even be a qol cause you can clear up space by allowing deconstruction of those projections).

So what we might perceive as a bug might actually be benificial for others,  We have to wait for players to create a convincing case on when these new interactions could be bugs.

**Comment by SyncIt21** [comment]
> While I agree a bitflag like this should not have a number of competing uses, I am not convinced that altering NO_DECONSTRUCT to have a different use was the best solution when there are a myriad of other less intrusive ones could have been chosen.

It's not ideal and I definitely have created a headache for a lot of people but it's not my fault. This flag has been abused for decades by allowing people to give their own definition and usages to it as I have explained in the PR body.

Somebody had to clean up this mess and I took the fall for it which is fine. In my PR body you can see why I agreed on giving this flag this final behavior, I'll just copy paste it here 

>as you have seen from my examples it is used in 3 cases
> 
> 1) Make an object disappear(doesn't dropping anything) when deconstructed
> 2) Stop it from being wrenched
> 3) Stop it from being deconstructed via tools

> We can't enforce points 2 & 3 inside deconstruct() which leaves us with only case 1) i.e. make the object disappear. And that's what i have done. Therefore after more than a decade or since this flag got introduced NO_DECONSTRUCT now has a new definition as of 2024
> 
> "Make an object disappear(don't dropping anything) when deconstructed either via tools or forcefully smashed apart"

I didn't choose this. The code base forced it upon me after misusing this flag for several years so it was the only 1 I was left with

**Pull request merged**